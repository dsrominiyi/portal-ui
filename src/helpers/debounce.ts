/**
 * Prevents an action from being called more that once every `wait` milliseconds
 * The initial triggering of the action will be delayed by `wait` milliseconds if `immediate` = false
 *
 * e.g. <button onClick={event => debounce(handleClick, 200)(event)} />
 */

let timeout: number | null;

const debounce = (action: (...args: any[]) => void, wait = 300, immediate = true) => (
  ...args: any[]
) => {
  const later = () => {
    timeout = null;
    if (!immediate) {
      action(...args);
    }
  };
  const callNow = immediate && !timeout;

  if (timeout) {
    clearTimeout(timeout);
  }

  timeout = (setTimeout as Window['setTimeout'])(later, wait);

  if (callNow) {
    action(...args);
  }
};

export default debounce;
