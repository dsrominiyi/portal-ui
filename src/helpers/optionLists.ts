export const maritalStatusOptions = [
  MaritalStatus.CommonLaw,
  MaritalStatus.Divorced,
  MaritalStatus.Married,
  MaritalStatus.Organisation,
  MaritalStatus.RegisteredCivilPartnership,
  MaritalStatus.Single,
  MaritalStatus.WidowWidower
];

export const consentTypeOptions = [
  MarketingConsentType.LegitimateInterest,
  MarketingConsentType.FaceToFace,
  MarketingConsentType.Phone,
  MarketingConsentType.Email,
  MarketingConsentType.Unsubscribe
];

export const contactMethodOptions = [
  ContactMethod.TextMessage,
  ContactMethod.Phone,
  ContactMethod.Email,
  ContactMethod.Post
];

export const additionalContactRelationshipOptions = [
  AdditionalContactRelationship.Spouse,
  AdditionalContactRelationship.Sibling,
  AdditionalContactRelationship.Parent,
  AdditionalContactRelationship.Child,
  AdditionalContactRelationship.Solicitor
];

export const contactTypeOptions = [
  ContactType.HomeNumber,
  ContactType.MobileNumber,
  ContactType.WorkNumber,
  ContactType.PersonalEmail,
  ContactType.WorkEmail
];

export const companyTypeStatusOptions = [
  CompanyTypeStatus.SoleProprietor,
  CompanyTypeStatus.Partnership,
  CompanyTypeStatus.Limited,
  CompanyTypeStatus.PublicLimited,
  CompanyTypeStatus.Other
];

export const otherCompanyTypeStatusOptions = [
  CompanyTypeStatus.AssociationClubSociety,
  CompanyTypeStatus.Charity,
  CompanyTypeStatus.Government,
  CompanyTypeStatus.LimitedLiabilityPartnership,
  CompanyTypeStatus.PartnershipLimited,
  CompanyTypeStatus.PartnershipOrdinary,
  CompanyTypeStatus.PrivateUnlimitedCompany,
  CompanyTypeStatus.TrustOrTrustees
];

export const referredByOptions = [
  'Practical Caravan',
  'Lead Linkedin',
  'C.V.R',
  'Autotrader Motorhome',
  'Yellow Pages',
  'Cross-selling',
  'Existing Business',
  'Lapsed Caravan',
  'Lapse CV',
  'Lapsed Home',
  'Ramsey Cricket Club',
  'Central Marketing',
  'Come Back Letter',
  'Fleet',
  'Stuarts Mailings',
  'Truck Tele',
  'Passing Trade',
  'Old Magazine Adverts',
  'Coachman Club',
  'Referral From NHF',
  'Recommended',
  'Vernon Property',
  'Insurecom Broker'
];

export const otherTitleOptions = [
  'Mx',
  'Master',
  'Brig.',
  'Bro.',
  'Capt.',
  'Chf.',
  'Col.',
  'Comdr.',
  'Comm.',
  'Cllr.',
  'Dame',
  'Dr.',
  'Exct.',
  'F/Lt.',
  'Gen.',
  'Hon.',
  'Judge',
  'Lady',
  'Lt.',
  'Lord',
  'Maj.',
  'Prof',
  'Rabbi',
  'Rev.',
  'Rt Rev',
  'Sir',
  'Sister',
  'Sq Ldr',
  'Rt Hon',
  'V. Rev',
  'W/Cdr'
];

export const individualAddressTypeOptions = ['Home', 'Correspondence'];

export const CompanyAddressTypeOptions = ['Correspondence', 'Trading', 'Registered'];

export const businessClassOptions = [
  'Accountancy',
  'Acoustic Engineer',
  'Actuarial Consultancy',
  'Acupuncture',
  'Addressing/Circularising Services',
  'Adjuster',
  'Advertising',
  'Aerial Erector',
  'Aerial Manufacturer',
  'Aerial Photography',
  'Aerial Supplier',
  'Aerial Survey',
  'Aerospace Industry',
  'Agricultural Engineer',
  'Agricultural Produce Distribution',
  'Agriculture',
  'Air Conditioning',
  'Air Transport'
];
