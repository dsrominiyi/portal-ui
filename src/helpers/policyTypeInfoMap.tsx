import React from 'react';

import PrivateCarIcon from '../assets/svg/icon-pt-private-car.svg';
import HouseholdIcon from '../assets/svg/icon-pt-household.svg';
import TravelIcon from '../assets/svg/icon-pt-travel.svg';
import CaravanIcon from '../assets/svg/icon-pt-caravan.svg';
import TruckIcon from '../assets/svg/icon-pt-truck.svg';
import CommercialCombinedIcon from '../assets/svg/icon-pt-commercial-combined.svg';
import PublicLiabilityIcon from '../assets/svg/icon-pt-public-liability.svg';
import CaterersIcon from '../assets/svg/icon-pt-caterers.svg';

export enum PolicyType {
  PrivateCar = 'pc',
  Household = 'hh',
  Travel = 'tv',
  Caravan = 'cn',
  Truck = 'cvt',
  CommercialCombined = 'cc',
  PublicLiability = 'pl',
  Caterers = 'ct'
}

export const policyTypeInfoMap = {
  [PolicyType.PrivateCar]: { icon: <PrivateCarIcon />, displayName: 'Private Car' },
  [PolicyType.Household]: { icon: <HouseholdIcon />, displayName: 'Household' },
  [PolicyType.Travel]: { icon: <TravelIcon />, displayName: 'Travel' },
  [PolicyType.Caravan]: { icon: <CaravanIcon />, displayName: 'Caravan' },
  [PolicyType.Truck]: { icon: <TruckIcon />, displayName: 'Truck' },
  [PolicyType.CommercialCombined]: {
    icon: <CommercialCombinedIcon />,
    displayName: 'Commercial Combined'
  },
  [PolicyType.PublicLiability]: {
    icon: <PublicLiabilityIcon />,
    displayName: 'Public Liability'
  },
  [PolicyType.Caterers]: { icon: <CaterersIcon />, displayName: 'Caterers' }
};
