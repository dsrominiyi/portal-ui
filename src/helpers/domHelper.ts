import debounce from './debounce';

export const preventBodyScroll = () => {
  const { scrollY } = window;
  document.body.style.position = 'fixed';
  document.body.style.top = document.body.style.top || `-${scrollY}px`;
};

export const allowBodyScroll = () => {
  const scrollY = document.body.style.top;
  document.body.style.position = '';
  document.body.style.top = '';
  return scrollY && window.scrollTo(0, -parseInt(scrollY, 10));
};

export const triggerOnEnter = (action: () => void, debounceWait?: number) => (
  event: React.KeyboardEvent<Element | HTMLElement>
) => {
  if (event.key === 'Enter') {
    debounce(action, debounceWait)();
  }
};
