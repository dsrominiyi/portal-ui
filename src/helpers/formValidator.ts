import { utc as moment } from 'moment';

import debounce from './debounce';
import { dateFieldValuesToDateString } from './utilFuncs';

/**
 * The formValidator helper functions supports a few different types of validation, these are
 * 'required', 'minLength', 'maxLength' and 'regex', where supplied values are simply used to
 * validate, additionally `predefinedRules` are made available, being made up of a range of rule
 * parts, these can then be imported for use in a form, and adjusted as individual needs require.
 */

export interface ValidationResult {
  isValid: boolean;
  message?: string;
}

export interface ValidationError {
  showError: boolean;
  message?: string;
}

export interface FormData {
  [formFieldKey: string]: FieldValue;
}

export enum RuleType {
  postcode = 'postcode'
}

export interface ValidationRule {
  fieldLabel: string;
  minLength?: number;
  maxLength?: number;
  required?: boolean;
  regex?: string | RegExp;
  isDate?: boolean;
  invalidValues?: string[];
  errorMessage?: string;
}

export interface ValidationRules {
  [formFieldKey: string]: ValidationRule;
}

export interface BatchValidationResult {
  [formFieldKey: string]: ValidationResult;
}

export interface FormErrors {
  [formFieldKey: string]: ValidationError;
}

export interface ValidationData {
  [formFieldKey: string]: ValidationResult;
}

export type ValidationStateArray = [
  BatchValidationResult,
  React.Dispatch<React.SetStateAction<BatchValidationResult>>
];

const isDateFieldValues = (value: FieldValue): value is DateFieldValues => {
  const objKeys = typeof value === 'object' && Object.keys(value as DateFieldValues);
  return objKeys && objKeys.includes('dd') && objKeys.includes('mm') && objKeys.includes('yyyy');
};

export const predefinedRules: ValidationRules = {
  postcode: {
    fieldLabel: 'postcode',
    minLength: 5,
    maxLength: 8,
    regex: /^[A-Z]{1,2}[0-9]{1,2}[A-Z]?(\s*[0-9][A-Z]{1,2})?$/,
    errorMessage: 'You must enter a valid UK postcode'
  },
  emailAddress: {
    fieldLabel: 'email address',
    maxLength: 254,
    regex: /^[^@]+@[^.]+\..+$/,
    errorMessage: 'You must enter a valid email address',
    required: true
  },
  phoneNumber: {
    fieldLabel: 'phone number',
    minLength: 7,
    maxLength: 50,
    regex: /^\+?[\d\s]+$/,
    errorMessage: 'You must enter a valid phone number',
    required: true
  },
  individualName: {
    fieldLabel: 'name',
    regex: /^[a-zA-Z-]+$/,
    minLength: 2,
    maxLength: 50
  },
  entityName: {
    fieldLabel: 'name',
    regex: /^[\w\s-]+$/,
    minLength: 2,
    maxLength: 50
  },
  registeredCoNumber: {
    fieldLabel: 'registration number',
    regex: /^([a-zA-Z]{2}[0-9]{6}|[0-9]{8})$/,
    errorMessage: 'You must enter a valid Companies House CRN',
    required: true
  }
};

export const validate = (rule: ValidationRule, value: FieldValue = ''): ValidationResult => {
  const {
    minLength,
    maxLength,
    required,
    regex,
    isDate,
    invalidValues,
    fieldLabel,
    errorMessage
  } = rule;

  const isBoolean = typeof value === 'boolean';
  const isArray = Array.isArray(value);
  const isEmptyArray = isArray && !(value as unknown[]).length;
  const nonRequiredAndEmpty = (!required && !value) || (!required && isEmptyArray);
  const requiredAndEmpty = (required && !value) || (required && isEmptyArray);

  if (isBoolean || nonRequiredAndEmpty) {
    return { isValid: true };
  }
  if (requiredAndEmpty) {
    return {
      isValid: false,
      message: errorMessage || `Please enter ${fieldLabel}`
    };
  }

  if (isDate) {
    const dateString = isDateFieldValues(value)
      ? dateFieldValuesToDateString(value)
      : (value as string);

    return (!dateString && !required) || moment(dateString, 'YYYY-MM-DD').isValid()
      ? { isValid: true }
      : {
          isValid: false,
          message: errorMessage || `${fieldLabel} should be a valid date`
        };
  }

  const stringValue = value as string;

  if (invalidValues && invalidValues.includes(stringValue)) {
    return {
      isValid: false,
      message:
        errorMessage || `${fieldLabel} cannot be one of the following: ${invalidValues.join(', ')}`
    };
  }
  if (minLength && minLength > stringValue.length) {
    return {
      isValid: false,
      message: errorMessage || `${fieldLabel} has a minimum length of ${minLength} characters`
    };
  }
  if (maxLength && maxLength < stringValue.length) {
    return {
      isValid: false,
      message: errorMessage || `${fieldLabel} has a maximum length of ${maxLength} characters`
    };
  }
  if (regex && !RegExp(regex).test(stringValue)) {
    return {
      isValid: false,
      message: errorMessage || `${fieldLabel} does not match the required format`
    };
  }

  return { isValid: true };
};

export const validateAll = (rules: ValidationRules, data: FormData): BatchValidationResult => {
  return Object.entries(data).reduce(
    (validationData, [formFieldKey, value]) => ({
      ...validationData,
      [formFieldKey]: validate(rules[formFieldKey], value)
    }),
    {}
  );
};

export const validateForm = (
  rules: ValidationRules,
  data: FormData,
  setValidationState: React.Dispatch<React.SetStateAction<BatchValidationResult>>
) => {
  const validationActions = () => {
    const validationResult = validateAll(rules, data);
    setValidationState(validationResult);
  };

  debounce(validationActions, 250, false)();
};

export const isFormValid = (validatedFields: BatchValidationResult) => {
  if (Object.keys(validatedFields).length === 0) {
    return false;
  }
  let allFieldsValid = true;

  for (const validatedField of Object.values(validatedFields)) {
    if (!validatedField.isValid) {
      allFieldsValid = false;
      break;
    }
  }

  return allFieldsValid;
};

export const buildErrors = (
  validationState: BatchValidationResult,
  submissionAttempted: boolean
): FormErrors => {
  return Object.entries(validationState).reduce(
    (errors, [formFieldKey, result]) => ({
      ...errors,
      [formFieldKey]: {
        showError: submissionAttempted && !result.isValid,
        message: result.message
      }
    }),
    {}
  );
};
