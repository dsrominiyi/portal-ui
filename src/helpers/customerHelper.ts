import { dateFieldValuesToDateString, dateStringToDateFieldValues } from './utilFuncs';
import { predefinedRules, ValidationRules } from './formValidator';

export const companyFormValidationRules: ValidationRules = {
  companyName: { ...predefinedRules.entityName, fieldLabel: 'company name', required: true },
  tradingName: { ...predefinedRules.entityName, fieldLabel: 'trading name' },
  typeStatus: { fieldLabel: 'company legal status', invalidValues: ['Other'], required: true },
  businessClass: { fieldLabel: 'business class', required: true },
  tradingSinceDate: { fieldLabel: 'trading since', isDate: true },
  parentCompany: { fieldLabel: 'is this company a subsidiary', required: true },
  parentCompanyName: { ...predefinedRules.entityName, fieldLabel: 'parent company name' },
  registeredCoNumber: predefinedRules.registeredCoNumber,
  contactDetails: { fieldLabel: 'contact details', required: true },
  addresses: { fieldLabel: 'addresses', required: true }
};

export const isIndividualCustomer = (customer: Customer): customer is Individual => {
  return typeof (customer as Individual).name !== 'undefined';
};

export const getIndividualFullName = ({ title, forename, middleName, surname }: IndividualName) => {
  return `${title} ${forename}${middleName && ` ${middleName}`} ${surname}`;
};

export const buildCompanyFromFormData = (data: NewCompanyData): Company => ({
  ...data,
  parentCompany: data.parentCompany || false,
  tradingSinceDate: dateFieldValuesToDateString(data.tradingSinceDate)
});

export const buildFormDataFromCompany = (data: Company): NewCompanyData => ({
  ...data,
  tradingSinceDate: data.tradingSinceDate
    ? dateStringToDateFieldValues(data.tradingSinceDate)
    : { dd: '', mm: '', yyyy: '' }
});
