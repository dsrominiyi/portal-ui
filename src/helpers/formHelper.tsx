import React, { ReactNode } from 'react';

import QuestionBox from '../components/QuestionBox';
import TextField, { Props as TextFieldProps } from '../components/TextField';
import AbiSearch, { Props as AbiSearchProps } from '../components/AbiSearch';
import Dropdown, { DropdownOption, Props as DropdownProps } from '../components/Dropdown';
import MultiChoice, { Props as MultiChoiceProps } from '../components/MultiChoice';
import DateFields, { Props as DateFieldsProps } from '../forms/common/DateFields';
import Lookup from '../components/Lookup';

import { ValidationError } from './formValidator';

import formStyles from '../forms/styles.scss';

interface InputConfig {
  type: InputType;
  label: string;
  buttonLabel?: string;
  secondaryButtonLabel?: string;
  value?: FieldValue;
  codeListId?: string;
  options?: DropdownOption[] | string[];
  attributes?: { [attr: string]: string };
  onChange?:
    | TextFieldProps['onChange']
    | AbiSearchProps['onChange']
    | DropdownProps['onChange']
    | DateFieldsProps['onChange']
    | MultiChoiceProps['onClick'];
  secondaryAction?: () => void;
}

export const buildInput = (config: InputConfig, fieldError?: ValidationError) => {
  let input: ReactNode;

  switch (config.type) {
    case InputType.Text:
      input = (
        <TextField
          key="0"
          additionalClasses={[formStyles.formTextField]}
          label={config.label}
          value={config.value as string}
          isInvalid={fieldError?.showError}
          onChange={config.onChange as TextFieldProps['onChange']}
          attributes={config.attributes}
        />
      );
      break;
    case InputType.Lookup:
      input = (
        <Lookup
          key="0"
          label={config.label}
          buttonLabel={config.buttonLabel}
          secondaryButtonLabel={config.secondaryButtonLabel}
          value={config.value as string}
          isInvalid={fieldError?.showError}
          onChange={config.onChange as TextFieldProps['onChange']}
          secondaryAction={config.secondaryAction}
          attributes={config.attributes}
        />
      );
      break;
    case InputType.AbiSearch:
      input = (
        <AbiSearch
          key="0"
          label={config.label}
          codeListId={config.codeListId || ''}
          onChange={config.onChange as AbiSearchProps['onChange']}
          attributes={config.attributes}
        />
      );
      break;
    case InputType.Dropdown:
      input = (
        <Dropdown
          key="0"
          label={config.label}
          options={config.options!}
          selected={config.value as string}
          isInvalid={fieldError?.showError}
          onChange={config.onChange as DropdownProps['onChange']}
          attributes={config.attributes}
        />
      );
      break;
    case InputType.DateFields:
      input = (
        <DateFields
          key="0"
          values={config.value as DateFieldValues}
          onChange={config.onChange as DateFieldsProps['onChange']}
          testId={config.attributes && config.attributes['test-id']}
        />
      );
      break;
    case InputType.MultiChoice:
      input = (
        <MultiChoice
          key="0"
          options={config.options as string[]}
          selectedOption={config.value as string}
          onClick={config.onChange as MultiChoiceProps['onClick']}
          attributes={config.attributes}
        />
      );
      break;
  }

  return (
    <QuestionBox
      key={config.label}
      text={[InputType.DateFields, InputType.MultiChoice].includes(config.type) ? config.label : ''}
      hasError={fieldError?.showError}
      errorText={fieldError?.message}
      inputFields={[input]}
    />
  );
};
