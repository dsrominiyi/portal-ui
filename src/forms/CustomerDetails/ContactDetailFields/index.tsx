import React, { FunctionComponent, Fragment } from 'react';

import QuestionBox from '../../../components/QuestionBox';
import Dropdown from '../../../components/Dropdown';
import TextField from '../../../components/TextField';

import { contactTypeOptions } from '../../../helpers/optionLists';
import { predefinedRules, ValidationError } from '../../../helpers/formValidator';

import formStyles from '../../styles.scss';

interface FieldErrors {
  type: ValidationError;
  detail: ValidationError;
}

export interface Props {
  type?: ContactType;
  detail: string;
  fieldErrors?: FieldErrors;
  onChange(values: Contact): void;
}

export const getDetailValidationRule = (type: ContactType) => {
  if ([ContactType.HomeNumber, ContactType.MobileNumber, ContactType.WorkNumber].includes(type)) {
    return predefinedRules.phoneNumber;
  }
  if ([ContactType.PersonalEmail, ContactType.WorkEmail].includes(type)) {
    return predefinedRules.emailAddress;
  }
  return { fieldLabel: 'contact detail', required: true };
};

export const ContactDetailFields: FunctionComponent<Props> = ({
  type = '' as ContactType,
  detail,
  fieldErrors = {} as FieldErrors,
  onChange
}: Props) => (
  <Fragment>
    <QuestionBox
      hasError={fieldErrors.type?.showError}
      errorText="Please confirm contact method"
      inputFields={[
        <Dropdown
          key="0"
          label="Contact method"
          options={contactTypeOptions}
          selected={type}
          isInvalid={fieldErrors.type?.showError}
          onChange={value => onChange({ type: value as ContactType, detail })}
          attributes={{ 'test-id': 'select-contact-method' }}
        />
      ]}
    />
    <QuestionBox
      hasError={fieldErrors.detail?.showError}
      errorText={fieldErrors.detail?.message}
      inputFields={[
        <TextField
          key="0"
          label="Contact"
          value={detail}
          additionalClasses={[formStyles.formTextField]}
          isInvalid={fieldErrors.detail?.showError}
          onChange={event => onChange({ type, detail: event.target.value })}
          attributes={{ 'test-id': 'contact-detail' }}
        />
      ]}
    />
  </Fragment>
);

ContactDetailFields.displayName = 'ContactDetailFields';
export default ContactDetailFields;
