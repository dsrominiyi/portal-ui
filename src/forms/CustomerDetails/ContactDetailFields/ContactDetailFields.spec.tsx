import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';

import ContactDetailFields, { Props, getDetailValidationRule } from '.';

import { predefinedRules } from '../../../helpers/formValidator';

let fields: ShallowWrapper<Props>;
let typeField: any;
let detailField: any;

const type = ContactType.MobileNumber;
const detail = '07040208066';

const onChange = jest.fn();
const fieldErrors = {
  type: { showError: true, message: 'error!' },
  detail: { showError: true, message: 'error!' }
};

const update = () => {
  [typeField, detailField] = (fields.props() as any).children;
};

describe('ContactDetailFields', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    fields = shallow(<ContactDetailFields detail="" onChange={onChange} />);
    update();
  });

  test('ContactDetailFields matches snapshot', () => {
    expect(fields).toMatchSnapshot();
  });

  test('ContactDetailFields with field errors matches snapshot', () => {
    fields.setProps({ fieldErrors });

    expect(fields).toMatchSnapshot();
  });

  test('selecting a type calls onChange with the updated value', () => {
    typeField.props.inputFields[0].props.onChange(type);

    expect(onChange).toHaveBeenCalledWith({ type, detail: '' });
  });

  test('entering the contact detail calls onChange with the updated value', () => {
    fields.setProps({ type });
    update();
    detailField.props.inputFields[0].props.onChange({ target: { value: detail } });

    expect(onChange).toHaveBeenCalledWith({ type, detail });
  });

  describe('getDetailValidationRule', () => {
    it('returns the phoneNumber validation rule for any number ContactType', () => {
      [ContactType.HomeNumber, ContactType.MobileNumber, ContactType.WorkNumber].forEach(
        numberType => {
          const rule = getDetailValidationRule(numberType);

          expect(rule).toEqual(predefinedRules.phoneNumber);
        }
      );
    });

    it('returns the emailAddress validation rule for any email ContactType', () => {
      [ContactType.PersonalEmail, ContactType.WorkEmail].forEach(emailType => {
        const rule = getDetailValidationRule(emailType);

        expect(rule).toEqual(predefinedRules.emailAddress);
      });
    });

    it('returns a default rule for any other type', () => {
      const rule = getDetailValidationRule('nonStandardType' as ContactType);

      expect(rule).toEqual({ fieldLabel: 'contact detail', required: true });
    });
  });
});
