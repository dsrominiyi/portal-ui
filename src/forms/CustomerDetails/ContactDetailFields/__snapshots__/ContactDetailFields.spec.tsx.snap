// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactDetailFields ContactDetailFields matches snapshot 1`] = `
<Fragment>
  <QuestionBox
    errorText="Please confirm contact method"
    inputFields={
      Array [
        <Dropdown
          attributes={
            Object {
              "test-id": "select-contact-method",
            }
          }
          label="Contact method"
          onChange={[Function]}
          options={
            Array [
              "Home Number",
              "Mobile Number",
              "Work Number",
              "Personal Email",
              "Work Email",
            ]
          }
          selected=""
        />,
      ]
    }
  />
  <QuestionBox
    inputFields={
      Array [
        <TextField
          additionalClasses={
            Array [
              "formTextField",
            ]
          }
          attributes={
            Object {
              "test-id": "contact-detail",
            }
          }
          label="Contact"
          onChange={[Function]}
          value=""
        />,
      ]
    }
  />
</Fragment>
`;

exports[`ContactDetailFields ContactDetailFields with field errors matches snapshot 1`] = `
<Fragment>
  <QuestionBox
    errorText="Please confirm contact method"
    hasError={true}
    inputFields={
      Array [
        <Dropdown
          attributes={
            Object {
              "test-id": "select-contact-method",
            }
          }
          isInvalid={true}
          label="Contact method"
          onChange={[Function]}
          options={
            Array [
              "Home Number",
              "Mobile Number",
              "Work Number",
              "Personal Email",
              "Work Email",
            ]
          }
          selected=""
        />,
      ]
    }
  />
  <QuestionBox
    errorText="error!"
    hasError={true}
    inputFields={
      Array [
        <TextField
          additionalClasses={
            Array [
              "formTextField",
            ]
          }
          attributes={
            Object {
              "test-id": "contact-detail",
            }
          }
          isInvalid={true}
          label="Contact"
          onChange={[Function]}
          value=""
        />,
      ]
    }
  />
</Fragment>
`;
